The error is:
```
Linking...
tests/system/.dub/build/application-debug-linux.posix-x86_64-dmd_2090-DB8FB99E8E2FF2C1325A116C295C574C/systemtest-simple.o:(.data.rel.ro+0x20): undefined reference to `_D5agora9consensus4data12PreimageInfo12__ModuleInfoZ'
tests/system/.dub/build/application-debug-linux.posix-x86_64-dmd_2090-DB8FB99E8E2FF2C1325A116C295C574C/systemtest-simple.o:(.data.rel.ro+0xa80): undefined reference to `_D5agora9consensus4data12PreimageInfo12__ModuleInfoZ'
collect2: error: ld returned 1 exit status
Error: linker exited with status 1
/home/circleci/dlang/dmd-2.090.1/linux/bin64/dmd failed with exit code 1.
```

It means one of your module is missing in `dub.json` (`agora.consensus.data.PreimageInfo`).

--- a/source/agora/api/Validator.d
+++ b/source/agora/api/Validator.d
@@ -105,20 +105,4 @@ public interface API : agora.api.FullNode.API
     ***************************************************************************/
 
     public void receivePreimage (PreimageInfo preimage);
-
-    /***************************************************************************
-
-        Checks if a pre-image exists
-
-        Params:
-            enroll_key = The key for the enrollment in which the pre-image is
-                contained.
-            height = The block height of the preimage to check existence
-
-        Returns:
-            true if the pre-image exists
-
-    ***************************************************************************/
-
-    public bool hasPreimage (Hash enroll_key, ulong height);
 }

--- a/source/agora/consensus/EnrollmentManager.d
+++ b/source/agora/consensus/EnrollmentManager.d
@@ -719,6 +719,43 @@ public class EnrollmentManager
 
     /***************************************************************************
 
+        Get validator's pre-image from the validator set
+
+        Params:
+            enroll_key = The key for the enrollment in which the pre-image is
+                contained.
+            result_image = will contain the PreimageInfo if exists
+
+        Returns:
+            true if getting pre-image is successfully processed
+
+    ***************************************************************************/
+
+    public bool getValidatorPreimage (const ref Hash enroll_key,
+        out PreimageInfo result_image) @trusted nothrow
+    {
+        try
+        {
+            auto results = this.db.execute("SELECT preimage from validator_set" ~
+                        " WHERE key = ?", enroll_key.toString());
+            if (!results.empty && results.oneValue!(byte[]).length != 0)
+            {
+                result_image =
+                    results.oneValue!(ubyte[]).deserializeFull!(PreimageInfo);
+            }
+        }
+        catch (Exception ex)
+        {
+            log.error("Exception occured, getValidatorPreimage:{}, exception:{}",
+                enroll_key, ex);
+            return false;
+        }
+
+        return true;
+    }
+
+    /***************************************************************************
+
         Add a pre-image information to a validator data
 
         Params:
@@ -1119,7 +1156,7 @@ unittest
     assert(validators[0].utxo_key == utxo_hash3);
 }
 
-/// tests for addPreimage and hasPreimage
+/// tests for addPreimage and getValidatorPreimage
 unittest
 {
     import agora.common.Amount;
@@ -1157,8 +1194,13 @@ unittest
     assert(man.add(0, &storage.findUTXO, enroll));
     assert(man.hasEnrollment(utxo_hash));
 
-    auto preimage = PreimageInfo(utxo_hash, enroll.random_seed, 1);
+    PreimageInfo result_image;
+    assert(man.getValidatorPreimage(utxo_hash, result_image));
+    assert(result_image == PreimageInfo.init);
+    auto preimage = PreimageInfo(utxo_hash, man.preimages[100], 1100);
     assert(man.addPreimage(preimage));
-    assert(man.hasPreimage(utxo_hash, 1));
-    assert(!man.hasPreimage(utxo_hash, 10));
+    assert(man.getValidatorPreimage(utxo_hash, result_image));
+    assert(result_image.enroll_key == utxo_hash);
+    assert(result_image.hash == man.preimages[100]);
+    assert(result_image.height == 1100);
 }

--- a/source/agora/node/Node.d
+++ b/source/agora/node/Node.d
@@ -497,12 +497,6 @@ public class Node : API
             this.network.sendPreimage(preimage);
     }
 
-    /// GET: /has_preimage
-    public bool hasPreimage (Hash enroll_key, ulong height)
-    {
-        return this.enroll_man.hasPreimage(enroll_key, height);
-    }
-
     /// GET: /get_preimage
     public PreimageInfo getPreimage (Hash enroll_key)
     {

--- a/source/agora/test/Base.d
+++ b/source/agora/test/Base.d
@@ -450,7 +450,7 @@ public interface TestAPI : API
     public abstract Enrollment createEnrollmentData();
 
     ///
-    public abstract PreimageInfo getPreimage (uint height);
+    public abstract void broadcastPreimage (uint height);
 
     ///
     public abstract void updateEnrolledHeight (Hash enroll_key, ulong height);
@@ -555,12 +555,12 @@ public class TestNode : Node, TestAPI
         return enroll;
     }
 
-    /// Get a node's own pre-image information used when revealing it
-    public override PreimageInfo getPreimage (uint height)
+    /// Broadcast a pre-image information to the network
+    public override void broadcastPreimage (uint height)
     {
         PreimageInfo preimage;
         this.enroll_man.getPreimage(height, preimage);
-        return preimage;
+        this.receivePreimage(preimage);
     }
 
     /// Set a enrolled height for the enrollment

--- a/source/agora/test/EnrollmentManager.d
+++ b/source/agora/test/EnrollmentManager.d
@@ -77,12 +77,11 @@ unittest
         retryFor(node.hasEnrollment(enroll.utxo_key) == true, 5.seconds));
 
     // tests for revealing a pre-image
-    auto preimage = node_1.getPreimage(1000);
-    node_1.receivePreimage(preimage);
+    node_1.broadcastPreimage(1000);
 
     // check if nodes contains the pre-image previously sent.
     nodes.each!(node =>
-        retryFor(node.hasPreimage(preimage.enroll_key, 999) == true, 5.seconds));
+        retryFor(node.hasPreimage(enroll.utxo_key, 999) == true, 5.seconds));
 }
 
 /// test for revealing a pre-image periodically

